### PR TITLE
RSDK-8146 - Close resources with done contexts or failed startup

### PR DESCRIPTION
**Pull request by cheukt**
Had a change in the modmanager where we call RemoveResource if an AddResource fails - reasoning

		// if there is an error while adding a modular resource, send a RemoveResource call to attempt clean up.
		// this way, the next time the robot attempts to add the resource, the module will hopefully have a clean slate.
		// This is to prevent the robot getting into a state where a resource exists on a module but because adding the resource
		// to the module causes an "already exists" error, the robot never learns of the correct state until restart.
		// A possible cause would be a resource that takes too long to start up, failing here but succeeding on the module.
		// NOTE: Calling RemoveResource only on those cases will not work as the resource wouldn't have been created yet.

but also might have unintended effects, so kept it out of this PR. This does mean only modules built with this fix (and hopefully when the python SDK and C++ SDK gets this https://viam.atlassian.net/browse/RSDK-8424) will be free of this issue

### CODE REVIEW

File: module/modmanager/manager.go [hunk 1]
```diff
@@ -488,6 +488,18 @@ func (mgr *Manager) addResource(ctx context.Context, conf resource.Config, deps
 
 	_, err = mod.client.AddResource(ctx, &pb.AddResourceRequest{Config: confProto, Dependencies: deps})
 	if err != nil {
+		// if there is an error while adding a modular resource, send a RemoveResource call to attempt clean up.
```
Comment (cheukt) [inline module/modmanager/manager.go:491]: still convincing myself that this would not be a regression, we could technically match on the specific error (`resource already exists`) but that feels more brittle. Doing this only on context done will not fix issues since the resource would not have been created yet

Comment (cheukt) [reply module/modmanager/manager.go]: removed the change since it feels too heavy handed - happy to discuss

File: module/module.go [hunk 2]
```diff
@@ -52,6 +52,9 @@ const (
 	// https://viam.atlassian.net/browse/RSDK-7521
 	// maxSupportedWebRTCTRacks is the max number of WebRTC tracks that can be supported given wihout hitting the sctp SDP message size limit.
 	maxSupportedWebRTCTRacks = 9
+
+	// NoModuleParentEnvVar indicates whether there is a parent for a module being started.
+	NoModuleParentEnvVar = "VIAM_NO_MODULE_PARENT"
```
Comment (maximpertsov) [inline module/module.go:57]: great change for the documentation alone

File: module/module_test.go [hunk 3]
```diff
@@ -723,6 +724,82 @@ func TestAttributeConversion(t *testing.T) {
...
+
+	m := setupLocalModule(t, ctx, logging.NewTestLogger(t))
+
+	var (
+		shouldCancel bool
+
+		constructCount int
+		closeCount     int
+	)
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (maximpertsov) [inline module/module_test.go:760]: minor, should we defer cancellation to ensure it gets called?

```suggestion
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
```

File: robot/impl/robot_reconfigure_test.go [hunk 4]
```diff
@@ -3819,7 +3819,7 @@ func TestResourceConstructCtxCancel(t *testing.T) {
 			conf resource.Config,
 			logger logging.Logger,
 		) (resource.Resource, error) {
-			contructCount++
+			constructCount++
 			wg.Add(1)
 			defer wg.Done()
 			cFunc.c()
```
Comment (maximpertsov) [inline robot/impl/robot_reconfigure_test.go:3825]: technically not part of this change, but why do we need to introduce a struct that wraps a cancellation function?

File: module/module.go [hunk 5]
```diff
@@ -467,6 +470,13 @@ func (m *Module) AddResource(ctx context.Context, req *pb.AddResourceRequest) (*
 		return nil, err
 	}
 
+	// If context has errored, even if construction succeeded we should close the resource and return the context error
+	// Use the shutdown context because otherwise any Close operations that rely on the context will immediately fail
+	if ctx.Err() != nil {
```
Comment (benjirewis) [inline module/module.go:475]: What deadline does this context carry? Do we have the plumbing set up to get the 1 minute timeout from the RDK on this context? I'm assuming logic like this might not be possible in C++ and Python?

File: module/module_test.go [hunk 6]
```diff
@@ -723,6 +724,82 @@ func TestAttributeConversion(t *testing.T) {
...
+		_, err := m.AddResource(ctx, &pb.AddResourceRequest{
+			Config: &v1.ComponentConfig{Name: "mymock", Api: shell.API.String(), Model: model.String()},
+		})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, constructCount, test.ShouldEqual, 1)
+		test.That(t, closeCount, test.ShouldEqual, 0)
+	})
+
+	t.Run("add resource but cancel the ctx during construction", func(t *testing.T) {
+		shouldCancel = true
```
Comment (benjirewis) [inline module/module_test.go:790]: [nit] I assume the order of these tests matter since they share `shouldCancel`? Should we just make `shouldCancel` a variable on the `MockConfig` instead? Same comment about new test in `robot_reconfigure_test.go`.

**Review by benjirewis — COMMENTED** [review]
Nice! Some qs.

File: module/modmanager/manager.go (hunk 1 above)
Comment (benjirewis) [reply module/modmanager/manager.go]: What issue could we run into calling `RemoveResource` if `AddResource` fails? An error from the module saying that the resource doesn't exist? Should we just `UncheckedError` it? I'm not sure I see what what you had was too heavy-handed.

Comment (cheukt) [reply module/modmanager/manager.go]: the main concern would probably be removing a resource that does exist - should the correct behavior in those cases be to remove then add, or just reconfigure?

Also might be a moot point if all the sdks handle cancels/timeouts properly

File: module/module.go (hunk 5 above)
Comment (cheukt) [reply module/module.go:475]: The deadline should be one minute, will update the comment to reference
from RDK to Python SDK, the plumbing exists (most likely for the C++ SDK as well). The deadline is carried over grpc, and should be available to the method.

File: module/modmanager/manager.go (hunk 1 above)
Comment (benjirewis) [reply module/modmanager/manager.go]: > should the correct behavior in those cases be to remove then add, or just reconfigure?

I think the only way we can say that a modular resource has "properly built" is if `AddResource` from the RDK returns no error. If `AddResource` failed for the RDK but succeeded for the module, I don't think the RDK would be in a position to handle reconfiguring the modular resource. In my mental model, it would make the most sense to remove, ignore and log any returned errors, then try adding again at a later point.

> Also might be a moot point if all the sdks handle cancels/timeouts properly

Agree. If SDK module libraries handle cancels/timeouts properly, and `close` their respective resources, I think `Remove` is redundant. I'd be fine if you don't want to use `RemoveResource` here and rely on https://viam.atlassian.net/browse/RSDK-8424 being implemented correctly and somewhat soon.

File: robot/impl/robot_reconfigure_test.go (hunk 4 above)
Comment (cheukt) [reply robot/impl/robot_reconfigure_test.go:3825]: no, went ahead and refactored the whole test

File: robot/impl/local_robot.go [hunk 7]
```diff
@@ -708,14 +708,25 @@ func (r *localRobot) newResource(
...
+	default:
 		return nil, errors.Errorf("invariant: no constructor for %q", conf.API)
 	}
-	return resInfo.DeprecatedRobotConstructor(ctx, r, conf, gNode.Logger())
+	if err != nil {
+		return nil, err
+	}
+
+	// If context has errored, even if construction succeeded we should close the resource and return the context error
+	// Use the shutdown context because otherwise any Close operations that rely on the context will immediately fail
```
[the remark this answers is not in the document]
Comment (cheukt) [reply robot/impl/local_robot.go]: done

File: module/module_test.go (hunk 3 above)
Comment (cheukt) [reply module/module_test.go]: done

File: module/modmanager/manager.go (hunk 1 above)
Comment (cheukt) [reply module/modmanager/manager.go]: ya, let's wait for more bug reports before fixing. I agree that in theory there shouldn't be much of an issue to add the additional Remove

File: module/module.go (hunk 5 above)
Comment (cheukt) [reply module/module.go:477]: added

File: module/module_test.go (hunk 6 above)
Comment (cheukt) [reply module/module_test.go]: refactored the whole set of tests so they should all be parallelizable

**Comment by cheukt** [comment]
manually tested with the case in https://viam.atlassian.net/browse/RSDK-8101

File: robot/impl/resource_manager.go [hunk 8]
```diff
@@ -614,6 +610,7 @@ func (manager *resourceManager) completeConfig(
 					defer func() {
 						stopSlowLogger()
 						resChan <- struct{}{}
+						lr.reconfigureWorkers.Done()
```
Comment (benjirewis) [inline robot/impl/resource_manager.go:613]: I'm a little suspicious of what `reconfigureWorkers` really represents. It looks like we're `Add`ing and `Done`ing twice for `processResource` calls if `!syncRes`. That seems wrong to me. It's only used in testing from what I can tell, so I won't block this PR on it; just noting that this caught my eye 👁️ .

**Review by benjirewis — APPROVED** [review]
Looks good! Thanks for small test refactor.

File: robot/impl/resource_manager.go (hunk 8 above)
Comment (cheukt) [reply robot/impl/resource_manager.go:613]: since reconfiguration is done level by level, the robot starts a goroutine to manage each separate resource configuration for each level (which also starts and waits for it's own goroutine but has the escape hatch of 1min) in parallel. it probably can be managed better by StoppableWorker haha, but it does make sense

Comment (benjirewis) [reply robot/impl/resource_manager.go:613]: Ah yeah if we're two levels of goroutine deep that does make sense. Forgot some of the nuances of this timeout system. `StoppableWorkers` might be good, but I'm more concerned with getting that struct into signaling for now.

**Review by maximpertsov — APPROVED** [review]
